reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py

```diff
@@ -0,0 +1,207 @@
+  # log_pytree_shape_and_statistics(
+  #     flax.jax_utils.unreplicate(model_params), workload.summary_writer)
+  return jax_utils.replicate(optimizer_state), opt_update_fn
+
+
+def l2_regularization(params, l2_decay_rank_threshold):
+  """Computes the squared l2 norm of the given parameters.
+
+  This function will only filter for parameters with
+  rank >= l2_decay_rank_threshold. So if this threshold is set to 2, then all
+  1d (and lower) parameter arrays, including all bias and batch norm params,
+  will be ignored in this computation.
+
+
+  Args:
+    params: Pytree containing parameters.
+    l2_decay_rank_threshold: The calculation will only include parameters with
+       param.ndim >= l2_decay_rank_threshold. Set to 2 to ignore all bias and
+       batch_norm params in the model.
+
+  Returns:
+    weight_l2: the squared l2 norm of all params matching the threshold.
+  """
+  weight_penalty_params = jax.tree_leaves(params)
+  weight_l2 = sum(
+      jnp.sum(x**2)
+      for x in weight_penalty_params
+      if x.ndim >= l2_decay_rank_threshold)
+  return weight_l2
+
+
+def _summary_str(param):
+  total_norm = jnp.linalg.norm(param.reshape(-1))
+  return '{} - {} - {}'.format(str(param.shape), param.size, total_norm)
+
+
+def log_pytree_shape_and_statistics(pytree, writer):
+  """Logs the shape and norm of every array in the pytree."""
+  if not pytree:
+    logging.info('Empty pytree')
+    return
+  logging.info('Printing model param shapes.')
+  shape_dict = jax.tree_map(_summary_str, pytree)
+
+  logging.info(shape_dict.pretty_repr())
+
+  total_params = jax.tree_util.tree_reduce(
+      operator.add, jax.tree_map(lambda x: x.size, pytree))
+  print('num params = ', total_params)
+
+
+def update_step(batch,
+                params,
+                batch_stats,
+                optimizer_state,
+                workload,
+                global_step,
+                hyperparameters,
+                opt_update_fn,
+                rng,
+                lr):
+
+  optimizer_state.hyperparams['learning_rate'] = lr
+
+  def _loss_fn(params):
+    """loss function used for training."""
+    params_rng, dropout_rng = jax.random.split(rng, 2)
+    (logits, logit_paddings), new_batch_stats = workload.model_fn(
+        params,
+        batch,
+        batch_stats,
+        spec.ForwardPassMode.TRAIN,
+        {'params' : params_rng, 'dropout' : dropout_rng})
+
+    loss = workload.loss_fn(batch['targets'], (logits, logit_paddings))
+    return loss, new_batch_stats
+
+  grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+  (loss, new_model_state), grad = grad_fn(params)
+
+  grad_clip = hyperparameters.grad_clip
+  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
+
+  grad_scaling_factor = grad_clip / (grad_norm + _GRAD_CLIP_EPS)
+  grad_scaling_factor = jax.lax.clamp(min=0.0, x=grad_scaling_factor, max=1.0)
+
+  grad = jax.tree_map(lambda x: x * grad_scaling_factor, grad)
+  loss, grad = lax.pmean((loss, grad), axis_name='batch')
```

Comment:
I don't remember what we settled on, but is it possible to pmean before calculating grad_norm? if not, can you add a comment saying we need pmean after otherwise the code hangs?

Reply:
so with batch size 256 doesn't matter if grad clip code is before / after pmean it still hangs 
I've commented out the grad_clip code and I'm unsure if this workload even needs gradient clipping , we never ran it without clip 

for now I've kept commented code and we can figure out if we need it / how to make it not hang if we do need it

Reply:
resolved in newest version of code, pmean happens before grad clipping now 
I've still eliminated jax.lax.cond in our grad clipping code as that's more efficient